meshmode/array_context.py

```diff
@@ -270,184 +285,222 @@ def get_container_context_recursively(ary):
-        return self._binary_op(operator.add, arg, self)
-
-    def __rsub__(self, arg):
-        return self._binary_op(operator.sub, arg, self)
-
-    def __rmul__(self, arg):
-        return self._binary_op(operator.mul, arg, self)
-
-    def __rtruediv__(self, arg):
-        return self._binary_op(operator.truediv, arg, self)
-
-    def __rpow__(self, arg):
-        return self._binary_op(operator.pow, arg, self)
-
-    def __rmod__(self, arg):
-        return self._binary_op(operator.mod, arg, self)
-
-    def __rdivmod__(self, arg):
-        return self._binary_op(divmod, arg, self)
-
-    def __pos__(self):
-        return self
-
-    def __neg__(self):
-        return self._unary_op(operator.neg, self)
-
-    def __abs__(self):
-        return self._unary_op(operator.abs, self)
-
-    # }}}
-
-    # {{{ comparison
-
-    def __eq__(self, arg):
-        return self._binary_op(self.array_context.np.equal, self, arg)
-
-    def __ne__(self, arg):
-        return self._binary_op(self.array_context.np.not_equal, self, arg)
-
-    def __lt__(self, arg):
-        return self._binary_op(self.array_context.np.less, self, arg)
-
-    def __gt__(self, arg):
-        return self._binary_op(self.array_context.np.greater, self, arg)
-
-    def __le__(self, arg):
-        return self._binary_op(self.array_context.np.less_equal, self, arg)
-
-    def __ge__(self, arg):
-        return self._binary_op(self.array_context.np.greater_equal, self, arg)
-
-    # }}}
-
-    # {{{ logical
-
-    def __and__(self, arg):
-        return self._binary_op(operator.and_, self, arg)
-
-    def __xor__(self, arg):
-        return self._binary_op(operator.xor, self, arg)
-
-    def __or__(self, arg):
-        return self._binary_op(operator.or_, self, arg)
-
-    def __rand__(self, arg):
-        return self._binary_op(operator.and_, arg, self)
-
-    def __rxor__(self, arg):
-        return self._binary_op(operator.xor, arg, self)
-
-    def __ror__(self, arg):
-        return self._binary_op(operator.or_, arg, self)
-
-    # }}}
-
-
-@get_container_context.register(ArrayContainerWithArithmetic)
-def _get_array_container_ac_arith(ary: ArrayContainerWithArithmetic):
-    return ary.array_context
+            strict_arg = ""
+
+        arg1_entry, arg1_container = arg1
+        arg2_entry, arg2_container = arg2
+        return (f"{op_str.format(arg1_entry, arg2_entry)} "
+                f"for {arg1_entry}, {arg2_entry} "
+                f"in zip({arg1_container}, {arg2_container}{strict_arg})")
+
+    elif isinstance(arg1, tuple):
+        arg1_entry, arg1_container = arg1
+        return (f"{op_str.format(arg1_entry, arg2)} "
+                f"for {arg1_entry} in {arg1_container}")
+
+    elif isinstance(arg2, tuple):
+        arg2_entry, arg2_container = arg2
+        return (f"{op_str.format(arg1, arg2_entry)} "
+                f"for {arg2_entry} in {arg2_container}")
+    else:
+        return op_str.format(arg1, arg2)
+
+
+def with_container_arithmetic(
```

Comment:
This isn't at all specific to array containers. It could conceivably live in pytools. What do you think?

Reply:
I don't really mind it staying in the `arraycontext` package when it splits off.

Reply:
OK, sounds good. I agree that that's also a reasonable place.